src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// //Ecryptes a message
+	pub fn push(
+		&mut self,
+		plaintext: &[u8],
+		ad: Option<&[u8]>,
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msglen = plaintext.len();
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
+		let mut poly = init(&OneTimeKey::from_slice(&block[..POLY1305_KEYSIZE])?);
+		use zeroize::Zeroize;
+		block.zeroize();
+		if adlen > 0 {
+			poly.update(ad.unwrap())?;
+		}
+		poly.update(&pad[..((16usize.overflowing_sub(adlen).0) & 15)])?;
+		block[0] = tag.bits();
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 1, &mut block)?;
+		poly.update(&block)?;
+		dst_out[0] = block[0];
+
+		chacha20_enc(
+			&self.get_key(),
+			&self.get_nonce(),
+			2,
+			plaintext,
+			&mut dst_out[cipherpos..],
+		)?;
+		poly.update(&dst_out[cipherpos..(cipherpos + msglen)])?;
+		poly.update(
+			&pad[..((16usize
+				.overflowing_sub(BLOCKSIZE)
+				.0
+				.overflowing_add(msglen)
+				.0) & 15)],
+		)?;
+		store_u64_into_le(&[adlen as u64], &mut slen);
+		poly.update(&slen)?;
+		store_u64_into_le(&[BLOCKSIZE as u64 + msglen as u64], &mut slen);
+		poly.update(&slen)?;
+		let mac = poly.finalize()?;
+		dst_out[macpos..(macpos + mac.get_length())].copy_from_slice(mac.unprotected_as_bytes());
+		xor_buf8(self.get_inonce_mut(), &dst_out[macpos..macpos + 8]);
+		increment4(self.get_counter_mut());
+		if tag == Tag::REKEY
+			|| secure_cmp(
+				self.get_counter_mut(),
+				&[0u8; SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES],
+			)
+			.is_ok()
+		{
+			self.rekey()?;
+		}
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// Decryptes a message
+	pub fn pull(
+		&mut self,
+		cipher: &[u8],
+		ad: Option<&[u8]>,
+		plaintext_out: &mut [u8],
+		tag_out: &mut Tag,
+	) -> Result<(), UnknownCryptoError> {
+		if cipher.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES {
+			return Err(UnknownCryptoError);
+		}
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let mlen = cipher.len() - SECRETSTREAM_XCHACHA20POLY1305_ABYTES;
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msgpos = core::mem::size_of::<Tag>();
+		let macpos = msgpos + mlen;
+		if plaintext_out.len() < mlen {
+			return Err(UnknownCryptoError);
+		}
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
```

Comment:
We can make `poly1305_key_gen` `pub(crate)` in `chacha20poly1305`, so that we avoid re-defining the key-generation multiple times.

Reply:
This is now available on master.